### PR TITLE
Allow attribute constraints to be relaxed during allocations.

### PR DESCRIPTION
**Pull request by spencerkimball**
This will be useful when up-replicating in order to achieve redundancy,
even if the placement isn't ideal.

### CODE REVIEW

File: storage/allocator.go [hunk 1]
```diff
@@ -155,46 +155,59 @@ func (a *allocator) capacityGossipUpdate(key string, _ bool) {
...
 }
 
 func (a *allocator) allocateTargetInternal(required proto.Attributes, existing []proto.Replica,
-	filter func(*proto.StoreDescriptor, *stat, *stat) bool) (*proto.StoreDescriptor, error) {
-	stores, sl, err := a.selectRandom(3, required, existing)
-	if err != nil {
-		return nil, err
-	}
+	relaxConstraints bool, filter func(*proto.StoreDescriptor, *stat, *stat) bool) (*proto.StoreDescriptor, error) {
+	attrs := append([]string(nil), required.Attrs...)
```
Comment (tamird) [inline storage/allocator.go]: is there any benefit to this over `append([]string{}, required.Attrs...)`?

File: storage/store.go [hunk 2]
```diff
@@ -1218,6 +1218,7 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
 	// We pushed the transaction(s) successfully, so resolve the intent(s).
 	for i, intent := range wiErr.Intents {
 		pushReply := bReply.Responses[i].GetValue().(*proto.InternalPushTxnResponse)
+		log.Infoc(ctx, fmt.Sprintf("push reply %d", i), log.Detail, pushReply)
```
Comment (tamird) [inline storage/store.go]: was this intended?

File: storage/allocator.go (hunk 1 above)
Comment (tamird) [inline storage/allocator.go]: also can probably avoid the copy if `relaxConstraints` is `false`

File: storage/allocator_test.go [hunk 3]
```diff
@@ -290,6 +290,52 @@ func TestAllocatorExistingReplica(t *testing.T) {
...
+	gossipStores(s.Gossip(), multiDCStores, t)
+
+	testCases := []struct {
+		required         []string // attribute strings
+		existing         []int    // existing store/node ID
+		relaxConstraints bool     // allow constraints to be relaxed?
+		expID            int      // expected store/node ID on allocate
+		expErr           bool
+	}{
+		{[]string{"a", "ssd"}, []int{}, true, 1, false},
```
Comment (tamird) [inline storage/allocator_test.go]: wow, this was really hard to figure out. A reference to `multiDCStores` would make this much easier to dereference

**Comment by tamird** [comment]
LGTM modulo nits

**Comment by bdarnell** [comment]
LGTM

File: storage/allocator.go (hunk 1 above)
Comment (spencerkimball) [inline storage/allocator.go]: I really don't know if it's faster. You could imagine that it doesn't do an allocation and []string{} does, though I doubt go is that stupid.

Not worried about the copy.

File: storage/allocator_test.go (hunk 3 above)
Comment (spencerkimball) [inline storage/allocator_test.go]: Yeah, good point. I addd a comment.

File: storage/store.go (hunk 2 above)
Comment (spencerkimball) [inline storage/store.go]: no, removing.

File: storage/allocator.go (hunk 1 above)
Comment (tamird) [inline storage/allocator.go]: @petermattis sez yours doesn't allocate and mine does, so now we know!

**Pull request merged**